Prototype of the performance benchmarking tool

shyamjvs opened the pull request:
[Design doc](https://docs.google.com/document/d/1olGQ7nHqoZVO714XtgBzrPvifR7nWu0pQWwelN3VQtM) for the tool.

Here's the basic benchmarking tool prototype designed as per the discussion we had and the design doc. To run the tool:
- `make` (in the benchmark directory)
- `./bin/benchmark` (for running the benchmark with default configuration, i.e. comparing jobs ci-kubernetes-e2e-gce-scalability and ci-kubernetes-kubemark-100-gce using their runs from last 48 hours with KS Test)
Note: `./bin/benchmark --help` (for knowing the flags)

cc @kubernetes/sig-scalability-misc @wojtek-t @gmarek

shyamjvs commented:
Removed Godeps and vendor code for now. I'll add them as a separate commit once this gets LGTM'ed.

gmarek commented on benchmark/cmd/benchmark.go:
```diff
@@ -0,0 +1,89 @@
+)
+
+var (
+	leftJobName, rightJobName	string
```
formatting.

shyamjvs replied:
Done.

gmarek commented on benchmark/pkg/comparer/comparer.go:
```diff
@@ -0,0 +1,35 @@
+// This is a wrapper function for various comparison schemes.
+func CompareJobsUsingScheme(jobComparisonData *util.JobComparisonData, scheme string, significanceLevel float64) error {
+	switch scheme {
+	case "K-S Test":
```
define constants for scheme names

shyamjvs replied:
Done.

gmarek commented on benchmark/pkg/comparer/schemes/kstest.go:
```diff
@@ -0,0 +1,46 @@
+
+	"k8s.io/perf-tests/benchmark/pkg/util"
+
+	"github.com/dgryski/go-onlinestats"
```
We need to make sure that the license of this packet is compatible with our.

shyamjvs replied:
Seems like it's a simple repo and doesn't have any license. Should we remove it?

gmarek replied:
I don't know:/ @fgrzadkowski

gmarek replied:
I think if there's no LICENSE file, we can't use this. You can write to the author, to add apache2 to the repo.

dgryski replied:
I will add an MIT license today or tomorrow.

shyamjvs replied:
Thanks for responding quickly.

shyamjvs replied:
Thanks to dgryski@ the license has been added now. Will update godep to vendor this newer code.

shyamjvs replied:
Done.

gmarek commented on benchmark/pkg/runselector/schemes/last_n_hours.go:
```diff
@@ -0,0 +1,49 @@
+	currentTime := uint64(time.Now().Unix())
+	for runNumber := latestRunNumber; runNumber > 0; runNumber-- {
+		if startStatus, err := googleGCSBucketUtils.CheckStartedStatus(job, runNumber); err == nil {
+			if (currentTime - startStatus.Timestamp) < uint64(3600 * n) {
```
give some slack for time skew

shyamjvs replied:
I don't understand why we need this slack for? We just want all tests that started within the last n hours (and finished of course).

gmarek replied:
Hour on one machine is not necesarily hour on another one, but I don't feel strongly about this one.

shyamjvs replied:
But I guess timestamp is absolute, right? Something like the number of seconds elapsed since the start of the universe (iirc 00:00:00 UTC, 1st Jan 1970)

gmarek replied:
Yes, but I don't see how this is related. Time skew is still a thing. But as I said - it's not a production system, so I don't care too much.

gmarek commented on benchmark/pkg/runselector/selector.go:
```diff
@@ -0,0 +1,35 @@
+// This is a wrapper function for various run selection schemes.
+func GetJobRunsUsingScheme(job string, scheme string, nValue int) ([]int, error) {
+	switch scheme {
+	case "last-n-runs":
```
constants

shyamjvs replied:
Done.

gmarek commented on benchmark/pkg/scraper/scraper.go:
```diff
@@ -0,0 +1,110 @@
+)
+
+const (
+	readingJunk     			= iota
```
formatting.

shyamjvs replied:
Done.

gmarek commented on benchmark/pkg/scraper/scraper.go:
```diff
@@ -0,0 +1,110 @@
+
+// This function takes a job name and a run number, gets that run's build log,
+// reads it and parses JSON summaries of API call and Pod startup latencies.
+func GetMetricsForRun(job string, run int) (map[string]*e2e.APIResponsiveness, map[string]*e2e.PodStartupLatency, error) {
```
I have a deja vu feeling - wasn't there a function like that somewhere?

shyamjvs replied:
Yes, you've got the right feeling. There was one such function in your compare tool (Ref: https://github.com/kubernetes/perf-tests/blob/master/compare/src/scraper.go#L39), however we aren't using it for 3 reasons:
- It doesn't work for kubemark logs where we don't have the metrics printed b/w those delimiters (used in the function)
- We want api and pod latencies, while that function gives logs size summary, resource usage summary and metrics summary (which includes api latencies but not pod startup latencies)
- This has more cleaner state machine for parsing which uniformly works for all the metrics we want

shyamjvs commented:
@gmarek Addressed your comments. Also, did some restructuring to make the code more cleaner and added the average-based test. PTAL while I work on adding unit tests.

gmarek commented on benchmark/Makefile:
```diff
@@ -0,0 +1,11 @@
+.PHONY: all benchmark clean
+
+OUT_DIR = bin
```
just inline `bin`

gmarek commented on benchmark/cmd/benchmark.go:
```diff
@@ -0,0 +1,89 @@
+	pflag.Parse()
+
+	// Select the runs of the left and right jobs to be used for comparison using the given run-selection scheme.
+	fmt.Printf("Selecting runs for jobs '%v' and '%v' using scheme '%v' with n = %v\n", leftJobName, rightJobName, runSelectionScheme, nValue)
```
glog here and everywhere below?

shyamjvs replied:
Done.

gmarek commented on benchmark/cmd/benchmark.go:
```diff
@@ -0,0 +1,89 @@
+	fs.StringVar(&rightJobName, "right-job-name", "ci-kubernetes-kubemark-100-gce", "Name of the job to be used as right hand side of comparison")
+	fs.StringVar(&runSelectionScheme, "run-selection-scheme", runselector.LastNHours, "Scheme for selecting the runs to be compared")
+	fs.IntVar(&nValue, "n", 24, "Value of 'n' (runs/hours depending on run-selection-scheme) to be used for run selection")
+	fs.StringVar(&comparisonScheme, "comparison-scheme", comparer.AvgTest, "Statistical test to be used as the algorithm for comparison")
```
List all known options.

shyamjvs replied:
Oops! Didn't notice the repo had a flags file. Done, thanks.

shyamjvs replied:
Done

gmarek commented on benchmark/pkg/runselector/schemes/last_n_runs.go:
```diff
@@ -0,0 +1,42 @@
+)
+
+// Returns a list of run numbers of last 'n' successfully completed runs of
+// 'job'. If it cannot find n runs, returns as many runs as it could collect.
```
I don't think that's a good behavior. It basically means that when the test starts to fail, we won't be able to use this. It should just take last n finished runs.

shyamjvs replied:
Good point. Changed it.

gmarek commented on benchmark/pkg/scraper/scraper.go:
```diff
@@ -0,0 +1,110 @@
+		case strings.Contains(line, testNameIdentifierTag):
+			testName = strings.Trim(strings.Split(line, testNameIdentifierTag)[1], " ")
+			buff.Reset()
+		case strings.Contains(line, apiLatencyResultTag):
```
Add check if you're in 'readingJunk' state. Same below.

shyamjvs replied:
Changed it.

gmarek commented on benchmark/pkg/scraper/scraper.go:
```diff
@@ -0,0 +1,110 @@
+	podStartupLatency := make(map[string]*e2e.PodStartupLatency)
+	testName := ""
+	state := readingJunk
+	for scanner.Scan() {
```
This is not really how state machines are implemented. It's more readable to switch on state, not on the line, as then you have clearly defined transitions.

shyamjvs replied:
True, it wasn't really a state machine (except for the usage of the term state) what I did there. However I've changed it to one now. Thanks!

gmarek commented on benchmark/pkg/scraper/scraper.go:
```diff
@@ -0,0 +1,110 @@
+	state := readingJunk
+	for scanner.Scan() {
+		line := scanner.Text()
+		switch {
```
I really don't like "free" switch statements. Please switch on some variable.

shyamjvs replied:
Done.

gmarek commented on benchmark/pkg/util/util.go:
```diff
@@ -0,0 +1,172 @@
+}
+
+// MetricComparisonData holds all the values corresponding to a metric's comparison.
+type MetricComparisonData struct {
```
Formatting. Run gofmt on this file.

shyamjvs replied:
Done (on all files).

gmarek commented on benchmark/pkg/util/util.go:
```diff
@@ -0,0 +1,172 @@
+func (j *JobComparisonData) PrettyPrint() {
+	const padding = 2
+	w := tabwriter.NewWriter(os.Stdout, 0, 0, padding, ' ', 0)
+	fmt.Fprintf(w, "E2E TEST\tVERB\tRESOURCE\tPERCENTILE\tMATCHED?\tCOMMENTS\n")
```
glog

shyamjvs replied:
I'm slightly hesitant in moving this logic from tabwriter to glog as there seems to be no good way in glog to print them as aligned columns.
However, I've changed it now to use tabwriter to write this output to a buffer and then calling glog to print that buffer. Hope that's fine.

gmarek replied:
Yup. I don't care what's written and how, but mostly where. And we should be consistent about that.

gmarek commented:
Please add tests for non-trivial functionality.

jeremyeder commented:
@jayunit100 @sjug

shyamjvs commented:
@gmarek I've added unit tests with good coverage. PTAL now.

gmarek commented on benchmark/pkg/scraper/scraper_test.go:
```diff
@@ -0,0 +1,112 @@
+)
+
+func TestGetMetricsForRun(t *testing.T) {
+	apiCallLatencies, podStartupLatency, err := GetMetricsForRun("ci-kubernetes-e2e-gce-scalability", 1937)
```
Run 1937 will disappear at some point. Please just download it, store it somewhere and run it locally.

shyamjvs replied:
Yes, I too don't like the fact that we rely on a particular real run's data. However, the problem is that `GetMetricsForRun` function calls `googleGCSBucketUtils.GetFileFromJenkinsGoogleBucket()` inside. I'm unaware of a way in which I can mock the function inside. How about I pick up the latest run (this way we also ensure that the unit test does work against the head of logs)?

gmarek replied:
Nope. It won't work as well, as it may be broken. Unit tests need to be isolated, so you shouldn't depend on network at all. There's no reason for testing network at all, only the logic behind it needs to be tested. If it's a problem, it means that logic is designed in a wrong way, as it should isolate reading data in a separate function.

shyamjvs replied:
Refactored the scraper code to separate the gcs calling part into a different function. Now it's using local test data. Made unit tests happy again!

gmarek commented on benchmark/pkg/util/util.go:
```diff
@@ -0,0 +1,175 @@
+}
+
+// PrettyPrint prints the job comparison data in a table form with columns aligned.
+func (j *JobComparisonData) PrettyPrint() {
```
No need for most of this logic - there's a `text/tabwriter` library.

shyamjvs replied:
I am already using that library. Did you mean something else?

gmarek replied:
Oh, sorry, I missed that line.

shyamjvs replied:
Done.

shyamjvs commented on benchmark/pkg/runselector/schemes/last_n_runs_test.go:
```diff
@@ -0,0 +1,36 @@
+	NumHours      = 3
+)
+
+func TestGetLastNJobRuns(t *testing.T) {
```
I'm sorry I wrongly named the variable as `NumHours` as part of copy-pasting. Changed it to (and what I meant was) `numRuns`. All we check is that the function can get the last 3 runs for that test (and has nothing to do with hours as this is the test for --last-n-runs selection scheme). For the --last-n-hours test, we check that we get at least one run in the last 24 hours for the job.

gmarek replied:
Right, but you also shouldn't depend on network, so you need to mock something.

shyamjvs commented on benchmark/pkg/comparer/schemes/avgtest_test.go:
```diff
@@ -0,0 +1,61 @@
+		},
+	}
+
+	CompareJobsUsingAvgTest(jobComparisonData, 0.4)
```
Done.

shyamjvs commented on benchmark/pkg/comparer/schemes/avgtest.go:
```diff
@@ -0,0 +1,48 @@
+				metricData.Matched = true
+			}
+		}
+		metricData.Comments += fmt.Sprintf("Avg_ratio=%.4f\t\tN_left=%v\tN_right=%v", leftRightAvgRatio, leftSampleCount, rightSampleCount)
```
There seems to be no good reason of using tabwriter here. It makes sense when we want to print multiple rows of text with columns that are tab-aligned. Here we just are generating a single line of comment (which doesn't know how the other comments are going to look like in order to align). Am I missing something here?

shyamjvs commented on benchmark/cmd/benchmark.go:
```diff
@@ -0,0 +1,91 @@
+	pflag.Parse()
+
+	// Select the runs of the left and right jobs to be used for comparison using the given run-selection scheme.
+	glog.Infof("Selecting runs for jobs '%v' and '%v' using scheme '%v' with n = %v\n", leftJobName, rightJobName, runSelectionScheme, nValue)
```
I changed the name of the variable to `nForRunSelection`. However, I guess the log itself seems clear enough. It looks like:

```
Selecting runs for jobs 'ci-kubernetes-e2e-gce-scalability' and 'ci-kubernetes-kubemark-100-gce' using scheme 'last-n-hours' with n = 24
```

The context for 'n' comes from the scheme name 'last-n-hours'. Do you still suggest changing it?

shyamjvs commented on benchmark/cmd/benchmark.go:
```diff
@@ -0,0 +1,91 @@
+	fs.Float64Var(&matchThreshold, "match-threshold", 0.8, "The threshold for metric comparison, interpretation depends on test used (significance level for KSTest, bound for ratio of avgs in AvgTest)")
+}
+
+func main() {
```
Done.

shyamjvs commented on benchmark/cmd/benchmark.go:
```diff
@@ -0,0 +1,91 @@
+	fs.StringVar(&leftJobName, "left-job-name", "ci-kubernetes-e2e-gce-scalability", "Name of the job to be used as left hand side of comparison")
+	fs.StringVar(&rightJobName, "right-job-name", "ci-kubernetes-kubemark-100-gce", "Name of the job to be used as right hand side of comparison")
+	fs.StringVar(&runSelectionScheme, "run-selection-scheme", runselector.LastNHours, "Scheme for selecting the runs to be compared")
+	fs.IntVar(&nValue, "n", 24, "Value of 'n' (runs/hours depending on run-selection-scheme) to be used for run selection")
```
Changed it to `n-for-run-selection`. Hope that's fine.

gmarek replied:
It shows that the design is wrong. You should just have two flags: `--last-runs-number` and `--last-runs-hours`, which are mutually exclusive.

shyamjvs replied:
Fixed this.

shyamjvs commented:
@gmarek Addressed the comments. PTAL

shyamjvs commented:
@gmarek Made the changes. Is it looking fine now?

gmarek commented:
The flags comment was not addressed.

gmarek commented on benchmark/pkg/runselector/schemes/last_n_runs_test.go:
```diff
@@ -0,0 +1,36 @@
+	numRuns       = 3
+)
+
+func TestGetLastNJobRuns(t *testing.T) {
```
This still depends on something on the network. I need to be able to run this test from my laptop when sitting in the airplane.

shyamjvs replied:
Fixed this now by mocking out all moving parts like network/file-reading .

shyamjvs commented:
@gmarek I've restructured the code now to use interfaces wherever they ought to be used. The `JobLogUtils` interface provides a set of functions that should together be sufficient for the metrics fetching part (run-selection + scraping) and it should be simple now to implement local mode too. Also unit tests have been isolated from all moving parts like network and file reads. PTAL

gmarek commented on benchmark/cmd/benchmark.go:
```diff
@@ -0,0 +1,129 @@
+		glog.Fatalf("Couldn't obtain log utils: %v", err)
+	}
+
+	nForRunSelection := nHoursCount
```
You should check if both are not set, and print error if they are.

shyamjvs replied:
Discussed it offline. Both are set due to default values. However we choose the relevant one based on the run-selection scheme.

gmarek commented:
Mostly LGTM - one nit, and it's good to go. Please make sure that it's still working after our recent test-infra changes.

gmarek commented:
```
./verify/verify-flags-underscore.py
Found flags in golang files not in the list of known flags. Please add these to verify/verify-flags/known-flags.txt
n-hours-count
n-runs-count
```

gmarek commented:
And please generate godeps.

shyamjvs commented:
Modified the flags file and added godeps in a separate commit.
Also, I've tested that it works with all those recent changes to our logs. Here's more info about the test run:

```
I0425 13:49:20.041906  191060 benchmark.go:65] Selecting runs for jobs 'ci-kubernetes-e2e-gce-scalability' and 'ci-kubernetes-kubemark-100-gce' using scheme 'last-n-hours' with n = 24
I0425 13:49:35.072024  191060 benchmark.go:71] Runs selected for job ci-kubernetes-e2e-gce-scalability: [2010 2009 2008 2007 2006 2005 2004 2003 2002 2001 2000 1999 1998 1997 1996 1995 1994 1993 1992 1991 1990]
I0425 13:49:50.402217  191060 benchmark.go:77] Runs selected for job ci-kubernetes-kubemark-100-gce: [2419 2418 2417 2416 2415 2414 2413 2412 2411 2410 2409 2408 2407 2406 2405 2404 2403 2402 2401 2400 2399 2398]
I0425 13:49:50.402263  191060 benchmark.go:89] Fetching metrics for the chosen runs of job ci-kubernetes-e2e-gce-scalability
I0425 13:49:58.313562  191060 benchmark.go:95] Fetching metrics for the chosen runs of job ci-kubernetes-kubemark-100-gce
I0425 13:50:06.449986  191060 benchmark.go:101] Flattening the metrics maps into per-metric structs
I0425 13:50:06.461166  191060 benchmark.go:108] Comparing metrics for the jobs using scheme 'Avg-Test' at a threshold value of 0.8
I0425 13:50:06.461919  191060 benchmark.go:127] Metric-wise results of comparison:
...
...
...
```

gmarek commented:
/lgtm
